src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/MSBuildNuGetProjectProvider.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.IO;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Utilities;
+using NuGet.Common;
+using NuGet.ProjectManagement;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    [Export(typeof(IProjectSystemProvider))]
+    [Name(nameof(MSBuildNuGetProjectProvider))]
+#if VS15
+    [Order(After = nameof(LegacyCSProjPackageReferenceProjectProvider))]
+#else
+    [Order(After = nameof(ProjectKNuGetProjectProvider))]
+#endif
+    public class MSBuildNuGetProjectProvider : IProjectSystemProvider
```

Comment:
>MSBuildNuGetProjectProvider [](start = 17, length = 27)

this should be split into multiple so that a provider only ever creates one type of project